packages/transformers/webextension/src/WebExtensionTransformer.js

```diff
@@ -267,91 +265,90 @@ async function collectDependencies(
-          );
-        }
-        asset.meta.webextBGInsert = program.background.scripts[0];
-      }
-    }
-  } else {
-    if (program.background?.service_worker) {
-      program.background.service_worker = asset.addURLDependency(
-        program.background.service_worker,
-        {
-          bundleBehavior: 'isolated',
-          loc: {
-            filePath,
-            ...getJSONSourceLocation(
-              ptrs['/background/service_worker'],
-              'value',
-            ),
-          },
-          env: {
-            context: 'service-worker',
-            sourceType:
-              program.background.type == 'module' ? 'module' : 'script',
-          },
-        },
-      );
-    }
-    if (hot) {
-      // Enable eval HMR for sandbox,
+    } else {
+      // Enable HMR for fetched localhost chunks
       const csp = program.content_security_policy || {};
       csp.extension_pages = cspPatchHMR(
         csp.extension_pages,
-        `http://${hmrOptions?.host || 'localhost'}`,
+        `http://${hmrOptions?.host || 'localhost'}:*`,
       );
       // Sandbox allows eval by default
       if (csp.sandbox) csp.sandbox = cspPatchHMR(csp.sandbox);
       program.content_security_policy = csp;
-      if (needRuntimeBG) {
-        if (!program.background) {
-          program.background = {};
-        }
-        if (!program.background.service_worker) {
-          program.background.service_worker = asset.addURLDependency(
-            './runtime/default-bg.js',
-            {
-              resolveFrom: __filename,
-              env: {context: 'service-worker'},
-            },
-          );
-        }
-        asset.meta.webextBGInsert = program.background.service_worker;
+    }
+
+    if (!program.background) {
+      program.background = {};
+    }
+
+    if (program.background.page) {
+      asset.meta.webextBGInsert = program.background.page;
+    } else if (isMV2 || program.background.scripts) {
+      if (!program.background.scripts) {
+        program.background.scripts = [];
       }
+      if (program.background.scripts.length == 0) {
+        program.background.scripts.push(
+          asset.addURLDependency('./runtime/default-bg.js', {
+            resolveFrom: __filename,
+          }),
+        );
+      }
+      asset.meta.webextBGInsert = program.background.scripts[0];
+    } else {
+      if (!program.background.service_worker) {
+        program.background.service_worker = asset.addURLDependency(
+          './runtime/default-bg.js',
+          {
+            resolveFrom: __filename,
+            env: {context: 'service-worker'},
+          },
+        );
+      }
+      asset.meta.webextBGInsert = program.background.service_worker;
+    }
+
+    if (!program.permissions) program.permissions = [];
+    if (!isMV2 && !program.permissions.includes('scripting')) {
+      program.permissions.push('scripting');
+    }
+    const hostPerms = [
+      ...new Set(program.content_scripts.flatMap(sc => sc.matches)),
```

Comment:
I'm getting an error here: `@parcel/transformer-webextension: Cannot read properties of undefined (reading 'flatMap')`

Trying to build a MV2 without content_scripts

Reply:
Already tracked in https://github.com/parcel-bundler/parcel/issues/9335